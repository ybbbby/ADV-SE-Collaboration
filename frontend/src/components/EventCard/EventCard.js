import React from 'react'
import PropTypes from 'prop-types'
import {
  Typography,
  Card,
  CardContent,
  CardActionArea,
  CardMedia,
  CardActions,
  Button,
  IconButton,
  Snackbar,
  Badge,
  Box,
} from '@material-ui/core'
import MuiAlert from '@material-ui/lab/Alert'
import FavoriteIcon from '@material-ui/icons/Favorite'
import ShareIcon from '@material-ui/icons/Share'
import { red } from '@material-ui/core/colors'
import { makeStyles } from '@material-ui/core/styles'
import { Link } from 'react-router-dom'
import ShareModal from '../ShareModal/ShareModal'

function Alert(props) {
  return <MuiAlert elevation={6} variant="filled" {...props} />
}

const useStyles = makeStyles(() => ({
  root: {
    boxShadow: '0 8px 24px 0 rgba(0,0,0,0.12)',
  },
  expand: {
    marginLeft: 'auto',
    textDecoration: 'none',
  },
  cardImage: {
    width: '100%',
    height: 0,
    paddingBottom: '56.25%',
    backgroundColor: 'rgba(0, 0, 0, 0.08)',
  },
  content: {
    margin: '-24% 16px 0',
    padding: '24px',
    position: 'relative',
    borderRadius: '4px',
    backgroundColor: '#fff',
    boxShadow:
      '0 2px 4px -2px rgba(0,0,0,0.24), 0 4px 24px -2px rgba(0, 0, 0, 0.2)',
  },
  title: {
    fontWeight: '800',
  },
}))

function EventCard(props) {
  const classes = useStyles()
  const { user, config, openLogin } = props
  const [like, setLike] = React.useState(config.liked)
  const [likeButtonColor, setLikeButtonColor] = React.useState(
    like ? red[500] : 'rgba(0, 0, 0, 0.54)'
  )
  const [alertOpen, setAlertOpen] = React.useState(false)
  const [shareModalOpen, setShareModalOpen] = React.useState(false)

  const clickLike = () => {
    if (!user) {
      openLogin(true)
      return
    }
    const currentStatus = like
    const color = currentStatus ? 'rgba(0, 0, 0, 0.54)' : red[500]
    const url = `/user/event/${config.id}/like`
    const requestForm = new FormData()
    requestForm.append('like', !currentStatus)
    fetch(url, {
      method: 'POST',
      body: requestForm,
    })
      .then((response) => {
        if (response.status !== 200) {
          throw Error(response.statusText)
        } else {
          return response.text()
        }
      })
      .then(() => {
        setLike(!currentStatus)
        setLikeButtonColor(color)
        setAlertOpen(true)
      })
      .catch((error) => {
        console.error(error)
      })
  }

  const closeAlert = (event, reason) => {
    if (reason === 'clickaway') {
      return
    }

    setAlertOpen(false)
  }

  const setOpenLogin = () => {
    if (!user) {
      openLogin(true)
    }
  }

  return (
    <>
      <Badge
        badgeContent={'host'}
        color="primary"
        invisible={!user || config.user_email !== user}
        style={{ display: 'block' }}
      >
        <Card className={classes.root}>
          <CardActionArea>
            <Link
              to={user ? `/event/${config.id}` : '#'}
              className={classes.expand}
            >
              <Box minHeight={250} position={'relative'}>
                <CardMedia className={classes.cardImage} image={config.image} />
                <CardContent className={classes.content}>
                  <Typography
                    variant="caption"
                    color="textSecondary"
                    component="p"
                  >
                    {config.time}
                  </Typography>
                  <Typography
                    gutterBottom
                    variant="h5"
                    component="h2"
                    color="textPrimary"
                    className={classes.title}
                  >
                    {config.name}
                  </Typography>

                  <Typography
                    variant="body1"
                    color="textSecondary"
                    component="p"
                  >
                    {config.address}
                  </Typography>
                </CardContent>
              </Box>
            </Link>
          </CardActionArea>
          <CardActions disableSpacing>
            <IconButton
              aria-label="add to favorites"
              onClick={() => clickLike()}
            >
              <FavoriteIcon style={{ color: likeButtonColor }} />
            </IconButton>
            <IconButton
              aria-label="share"
              onClick={() => setShareModalOpen(true)}
            >
              <ShareIcon />
            </IconButton>
            <Link
              to={user ? `/event/${config.id}` : '#'}
              className={classes.expand}
            >
              <Button
                size="small"
                color="primary"
                onClick={() => setOpenLogin()}
              >
                Learn More
              </Button>
            </Link>
          </CardActions>
        </Card>
      </Badge>
      <ShareModal
        url={`localhost:2000/event/${config.id}`}
        handleClose={() => setShareModalOpen(false)}
        open={shareModalOpen}
      />
      <Snackbar
        open={alertOpen}
        autoHideDuration={3000}
        onClose={() => closeAlert()}
      >
        {like ? (
          <Alert onClose={() => closeAlert()} severity="success">
            Saved to your favourite events!
          </Alert>
        ) : (
          <Alert onClose={() => closeAlert()} severity="info">
            Remove this event from your favourite events!
          </Alert>
        )}
      </Snackbar>
    </>
  )
}

EventCard.propTypes = {
  user: PropTypes.string.isRequired,
  config: PropTypes.object.isRequired,
  openLogin: PropTypes.bool,
}

export default EventCard
