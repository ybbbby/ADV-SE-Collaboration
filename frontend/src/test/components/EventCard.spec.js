/* eslint-env jest */
import { shallow } from 'enzyme'
import { HashRouter as Router } from 'react-router-dom'
import React from 'react'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import EventCard from '../../components/EventCard/EventCard'

const eventConfig = {
  address: '1275 York Avenue, New York, NY, USA',
  attended: false,
  author: 'test user',
  category: 'music',
  comments: [],
  description: '22211',
  event_id: '160685996825045',
  image: 'https://yesok.s3.amazonaws.com/uploads%2FIMG_2396.JPG',
  latitude: '40.764215',
  liked: false,
  longitude: '-73.956104',
  name: 'event name',
  num_likes: 0,
  time: '2020-12-25 17:49:00',
  user_email: 'test@columbia.edu',
  zipcode: '10025',
}

const openLogin = jest.fn()

describe('EventCard', () => {
  test('renders event info correctly when logged in', () => {
    render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    expect(
      screen.queryByText('1275 York Avenue, New York, NY, USA')
    ).toBeDefined()
    expect(screen.queryByText('event name')).toBeDefined()
    expect(screen.queryByText('music')).toBeDefined()
    expect(screen.queryByText('test user')).toBeDefined()
  })

  test('judge if category exists', () => {
    eventConfig.category = ''
    render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    expect(screen.queryByText('music')).toBeNull()
  })

  test('host badge invisible when the user is not host', () => {
    render(
      <Router>
        <EventCard
          user={'other user'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    expect(screen.getByText('host').closest('span')).toHaveClass(
      'MuiBadge-invisible'
    )
  })

  test('clicks like button when logged in', async () => {
    fetch.mockResponseOnce({ status: 200 })
    const { getByText } = render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(
      screen.getByLabelText('add to favorites', { selector: 'button' })
    )
    await waitFor(() => {
      expect(getByText('Saved to your favourite events!')).toBeInTheDocument()
    })
  })

  test('clicks like button when logged in (fail)', async () => {
    fetch.mockReject(() => Promise.reject('API is down'))
    const { queryByText } = render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(
      screen.getByLabelText('add to favorites', { selector: 'button' })
    )
    await waitFor(() => {
      expect(queryByText('Saved to your favourite events!')).toBeNull()
    })
  })

  test('cancel like when logged in', async () => {
    fetch.mockResponseOnce({ status: 200 })
    eventConfig.liked = true
    const { getByText } = render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(
      screen.getByLabelText('add to favorites', { selector: 'button' })
    )
    await waitFor(() => {
      expect(
        getByText('Remove this event from your favourite events!')
      ).toBeInTheDocument()
    })
  })

  test('clicks like button when not logged in', () => {
    render(
      <Router>
        <EventCard user={null} config={eventConfig} openLogin={openLogin} />
      </Router>
    )
    fireEvent.click(
      screen.getByLabelText('add to favorites', { selector: 'button' })
    )
    expect(screen.findAllByRole('presentation')).toBeDefined()
  })

  test('clicks learn more when logged in', () => {
    render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(screen.getByRole('button', { name: 'Learn More' }))
    expect(screen.queryByText('Welcome Back!')).toBeNull()
  })

  test('clicks learn more when not logged in', () => {
    render(
      <Router>
        <EventCard user={null} config={eventConfig} openLogin={openLogin} />
      </Router>
    )
    fireEvent.click(screen.getByRole('button', { name: 'Learn More' }))
    expect(screen.findAllByRole('presentation')).toBeDefined()
  })

  test('clicks share button when logged in', async () => {
    render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(screen.getByLabelText('share', { selector: 'button' }))
    expect(screen.getByText('Share the event')).toBeInTheDocument()
  })

  test('close alert after clicking the like button', async () => {
    fetch.mockResponseOnce({ status: 200 })
    const { getByRole, queryByText } = render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(
      screen.getByLabelText('add to favorites', { selector: 'button' })
    )
    await waitFor(() => {
      expect(getByRole('button', { name: 'Close' })).toBeInTheDocument()
    })
    fireEvent.click(screen.getByRole('button', { name: 'Close' }))
    await waitFor(() => {
      expect(queryByText('Saved to your favourite events!')).toBeNull()
    })
  })

  test('close alert after canceling like', async () => {
    fetch.mockResponseOnce({ status: 200 })
    eventConfig.liked = true
    const { getByRole, queryByText } = render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(
      screen.getByLabelText('add to favorites', { selector: 'button' })
    )
    await waitFor(() => {
      expect(getByRole('button', { name: 'Close' })).toBeInTheDocument()
    })
    fireEvent.click(screen.getByRole('button', { name: 'Close' }))
    await waitFor(() => {
      expect(
        queryByText('Remove this event from your favourite events!')
      ).toBeNull()
    })
  })

  test('close alert after clicking away', async () => {
    fetch.mockResponseOnce({ status: 200 })
    const { getByRole, queryByText } = render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(
      screen.getByLabelText('add to favorites', { selector: 'button' })
    )
    await waitFor(() => {
      expect(getByRole('button', { name: 'Close' })).toBeInTheDocument()
    })
    fireEvent.click(
      screen.getByLabelText('add to favorites', { selector: 'button' })
    )
    await waitFor(() => {
      expect(queryByText('Saved to your favourite events!')).toBeNull()
    })
  })

  test('close share modal when clicks the close button', async () => {
    render(
      <Router>
        <EventCard
          user={'test@columbia.edu'}
          config={eventConfig}
          openLogin={openLogin}
        />
      </Router>
    )
    fireEvent.click(screen.getByLabelText('share', { selector: 'button' }))
    fireEvent.click(screen.getByRole('button', { name: 'Cancel' }))
    // the handleClose function is tested in the ShareModal tests
  })

  it('snapshot when logged in', () => {
    const wrapper = shallow(
      <EventCard
        user={'test@columbia.edu'}
        config={eventConfig}
        openLogin={openLogin}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('snapshot when not logged in', () => {
    const wrapper = shallow(
      <EventCard user={null} config={eventConfig} openLogin={openLogin} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})
