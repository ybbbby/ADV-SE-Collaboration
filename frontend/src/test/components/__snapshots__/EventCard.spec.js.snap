// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventCard snapshot when logged in 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Badge))
    badgeContent="host"
    color="primary"
    invisible={false}
    style={
      Object {
        "display": "block",
      }
    }
  >
    <WithStyles(ForwardRef(Card))
      className="makeStyles-root-169"
    >
      <WithStyles(ForwardRef(CardActionArea))>
        <Link
          className="makeStyles-expand-170"
          to="/event/160685996825045"
        >
          <Styled(MuiBox)
            minHeight={250}
            position="relative"
          >
            <WithStyles(ForwardRef(CardMedia))
              className="makeStyles-cardImage-171"
              image="https://yesok.s3.amazonaws.com/uploads%2FIMG_2396.JPG"
            />
            <WithStyles(ForwardRef(CardContent))
              className="makeStyles-content-172"
            >
              <WithStyles(ForwardRef(Typography))
                color="textSecondary"
                component="p"
                variant="caption"
              >
                2020-12-25 17:49:00
              </WithStyles(ForwardRef(Typography))>
              <WithStyles(ForwardRef(Typography))
                className="makeStyles-title-173"
                color="textPrimary"
                component="h2"
                gutterBottom={true}
                variant="h5"
              >
                event name
              </WithStyles(ForwardRef(Typography))>
              <WithStyles(ForwardRef(Typography))
                color="textSecondary"
                component="p"
                variant="body1"
              >
                1275 York Avenue, New York, NY, USA
              </WithStyles(ForwardRef(Typography))>
            </WithStyles(ForwardRef(CardContent))>
          </Styled(MuiBox)>
        </Link>
      </WithStyles(ForwardRef(CardActionArea))>
      <WithStyles(ForwardRef(CardActions))
        disableSpacing={true}
      >
        <WithStyles(ForwardRef(IconButton))
          aria-label="add to favorites"
          onClick={[Function]}
        >
          <Memo([object Object])
            style={
              Object {
                "color": "#f44336",
              }
            }
          />
        </WithStyles(ForwardRef(IconButton))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="share"
          onClick={[Function]}
        >
          <Memo([object Object]) />
        </WithStyles(ForwardRef(IconButton))>
        <Link
          className="makeStyles-expand-170"
          to="/event/160685996825045"
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            onClick={[Function]}
            size="small"
          >
            Learn More
          </WithStyles(ForwardRef(Button))>
        </Link>
      </WithStyles(ForwardRef(CardActions))>
    </WithStyles(ForwardRef(Card))>
  </WithStyles(ForwardRef(Badge))>
  <ShareModal
    handleClose={[Function]}
    open={false}
    url="http://yes-ok.herokuapp.com/#/event/160685996825045"
  />
  <WithStyles(ForwardRef(Snackbar))
    autoHideDuration={3000}
    onClose={[Function]}
    open={false}
  >
    <Alert
      onClose={[Function]}
      severity="success"
    >
      Saved to your favourite events!
    </Alert>
  </WithStyles(ForwardRef(Snackbar))>
</Fragment>
`;

exports[`EventCard snapshot when not logged in 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Badge))
    badgeContent="host"
    color="primary"
    invisible={true}
    style={
      Object {
        "display": "block",
      }
    }
  >
    <WithStyles(ForwardRef(Card))
      className="makeStyles-root-169"
    >
      <WithStyles(ForwardRef(CardActionArea))>
        <Link
          className="makeStyles-expand-170"
          to="#"
        >
          <Styled(MuiBox)
            minHeight={250}
            position="relative"
          >
            <WithStyles(ForwardRef(CardMedia))
              className="makeStyles-cardImage-171"
              image="https://yesok.s3.amazonaws.com/uploads%2FIMG_2396.JPG"
            />
            <WithStyles(ForwardRef(CardContent))
              className="makeStyles-content-172"
            >
              <WithStyles(ForwardRef(Typography))
                color="textSecondary"
                component="p"
                variant="caption"
              >
                2020-12-25 17:49:00
              </WithStyles(ForwardRef(Typography))>
              <WithStyles(ForwardRef(Typography))
                className="makeStyles-title-173"
                color="textPrimary"
                component="h2"
                gutterBottom={true}
                variant="h5"
              >
                event name
              </WithStyles(ForwardRef(Typography))>
              <WithStyles(ForwardRef(Typography))
                color="textSecondary"
                component="p"
                variant="body1"
              >
                1275 York Avenue, New York, NY, USA
              </WithStyles(ForwardRef(Typography))>
            </WithStyles(ForwardRef(CardContent))>
          </Styled(MuiBox)>
        </Link>
      </WithStyles(ForwardRef(CardActionArea))>
      <WithStyles(ForwardRef(CardActions))
        disableSpacing={true}
      >
        <WithStyles(ForwardRef(IconButton))
          aria-label="add to favorites"
          onClick={[Function]}
        >
          <Memo([object Object])
            style={
              Object {
                "color": "#f44336",
              }
            }
          />
        </WithStyles(ForwardRef(IconButton))>
        <WithStyles(ForwardRef(IconButton))
          aria-label="share"
          onClick={[Function]}
        >
          <Memo([object Object]) />
        </WithStyles(ForwardRef(IconButton))>
        <Link
          className="makeStyles-expand-170"
          to="#"
        >
          <WithStyles(ForwardRef(Button))
            color="primary"
            onClick={[Function]}
            size="small"
          >
            Learn More
          </WithStyles(ForwardRef(Button))>
        </Link>
      </WithStyles(ForwardRef(CardActions))>
    </WithStyles(ForwardRef(Card))>
  </WithStyles(ForwardRef(Badge))>
  <ShareModal
    handleClose={[Function]}
    open={false}
    url="http://yes-ok.herokuapp.com/#/event/160685996825045"
  />
  <WithStyles(ForwardRef(Snackbar))
    autoHideDuration={3000}
    onClose={[Function]}
    open={false}
  >
    <Alert
      onClose={[Function]}
      severity="success"
    >
      Saved to your favourite events!
    </Alert>
  </WithStyles(ForwardRef(Snackbar))>
</Fragment>
`;
